silo-vaults/contracts/MetaMorpho.sol

```diff
@@ -738,59 +717,57 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
 
     /// @dev Sets the cap of the market defined by `id` to `supplyCap`.
     /// @dev Assumes that the inputs `marketParams` and `id` match.
-    function _setCap(MarketParams memory marketParams, Id id, uint184 supplyCap) internal {
-        MarketConfig storage marketConfig = config[id];
+    function _setCap(IERC4626 market, uint184 supplyCap) internal virtual {
+        MarketConfig storage marketConfig = config[market];
 
         if (supplyCap > 0) {
             if (!marketConfig.enabled) {
-                withdrawQueue.push(id);
+                withdrawQueue.push(market);
 
                 if (withdrawQueue.length > ConstantsLib.MAX_QUEUE_LENGTH) revert ErrorsLib.MaxQueueLengthExceeded();
 
                 marketConfig.enabled = true;
 
                 // Take into account assets of the new market without applying a fee.
-                _updateLastTotalAssets(lastTotalAssets + MORPHO.expectedSupplyAssets(marketParams, address(this)));
+                _updateLastTotalAssets(lastTotalAssets + _expectedSupplyAssets(market, address(this)));
 
                 emit EventsLib.SetWithdrawQueue(msg.sender, withdrawQueue);
             }
 
             marketConfig.removableAt = 0;
+            IERC20(asset()).approve(address(market), 0);
         }
 
         marketConfig.cap = supplyCap;
+        // TODO should we use forceApprove? and should we use max?
```

Comment:
Approve, and max looks good to me. I see no way to predict the exact amount to be deposited.
```suggestion
```

Reply:
I removed comment